docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+type PeerFailure struct {
+  lastFailure time.Time
+  numFailures int8
+}
+
+type LRUTxCache struct {
+// Keeping track of invalid transactions within the cache ;
+  invalidCachedTx map[types.TxKey]bool
+}
+
+type CListMempool struct {
+ // ..existing fields
+  peerFailureMap map[nodeID]*PeerFailure
+
+}
+```
+
+>mempool/v0/clist_mempool.go#L239
+```golang
+
+if !mem.cache.Push(tx) { // if the transaction already exists in the cache
+		// Record a new sender for a tx we've already seen.
+		// Note it's possible a tx is still in the cache but no longer in the mempool
+		// (eg. after committing a block, txs are removed from mempool but not cache),
+		// so we only record the sender for txs still in the mempool.
+		if e, ok := mem.txsMap.Load(tx.Key()); ok {
+			memTx := e.(*clist.CElement).Value.(*mempoolTx)
+			memTx.senders.LoadOrStore(txInfo.SenderID, true)
+			// TODO: consider punishing peer for dups,
+			// its non-trivial since invalid txs can become valid,
+			// but they can spam the same tx with little cost to them atm.
+		}
+
+    // If transaction was invalid, we need to remember the peer information
+    if _, ok := mem.cache.invalidCachedTx.Load(tx.Key); ok {
+        mem.banPeer(peerID)  
+    }
+		return mempool.ErrTxInCache
+	}
+
+```
+
+
+```golang
+
+    func (mem* ClistMempool) banPeer(peerID NodeID) {
+        numFails := 0
+         // Check whether this peer has sent us transactions that fail
+        if val, ok := mem.peerFailureMap[peerID]; ok {
+         lastFailureT := val.lastFailure
+         numFails = val.numFails
+        // if the failure was recent enough, update the number of failures and 
+        // ban peer if applicable
+        if time.Since(lastFailureT) <= failureResetInterval {
+          if numFails == maxAllowedFailures - 1 {
+              // Send Ban request to p2p 
+            }
+          }
+        }
+        // Update the time of the last failure     
+        mem.peerFailureMap[peerID] = { time.Now(), numFailures + 1}
+    }
+```
+If transactions with `ResponseCheckTx.code > 1` are not deleted from the cache and we want to ban them on the first offence,
+we can skip the second `if` in the code above but call `banPeer` immediately at the end of the function. The function `banPeer` should simply forward the `peerID` to the p2p layer. In fact, we do not need to store any information on this peer as the node will remove it from its peer set.
+
+**Signaling the ban to p2p**
+
+As it is the mempool **reactor** that has access to the p2p layer, not the actual mempool implementation, the peer banning function will most likely have to send the peer ID to a channel to inform the reactor of the fact that this peer should be banned. This would require adding a channel for communication into the `CListMempool` on construction, and adding a routine in the mempool ractor that waits on a response via this channel. However, the actual implementation of this is yet to be defined.
```

Comment:
```suggestion
As it is the mempool **reactor** that has access to the p2p layer, not the actual mempool implementation, the peer banning function will most likely have to send the peer ID to a channel to inform the reactor of the fact that this peer should be banned. This would require adding a channel for communication into the `CListMempool` on construction, and adding a routine in the mempool reactor that waits on a response via this channel. However, the actual implementation of this is yet to be defined.
```